{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d2cebff7",
   "metadata": {},
   "source": [
    "http://www.biopax.org/\n",
    "Biological Pathway Exchange (BioPAX) is a standard language that aims to enable integration, exchange, visualization and analysis of biological pathway data. Specifically, BioPAX supports data exchange between pathway data groups and thus reduces the complexity of interchange between data formats by providing an accepted standard format for pathway data. It is an open and collaborative effort by the community of researchers, software developers, and institutions. BioPAX is defined in OWL DL and is represented in the RDF/XML format. For more details, see Demir E et al. 2010. The BioPAX community standard for pathway data sharing, Nature Biotechnology. 28(9).\n",
    "Specification: http://www.biopax.org/release/biopax-level3-documentation.pdf\n",
    "https://pypi.org/project/pybiopax/\n",
    "https://github.com/indralab/pybiopax\n",
    "For the reactions you need probably the `BiochemicalReaction` and `Degradation`.\n",
    "For the PhysicalEntities you would use `Protein` and `RNA`.\n",
    "For the inhibition you will probably use `Modulation`.\n",
    "Have a look at the specification and the introduction to BioPax.\n",
    "Example for fetching information\n",
    "https://github.com/indralab/pybiopax/blob/master/notebooks/tutorial.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f85ebb03",
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import pybiopax\n",
    "from pathlib import Path\n",
    "\n",
    "# from combine_notebooks import RESOURCES_DIR\n",
    "from pybiopax.biopax import UnificationXref, Control, CellularLocationVocabulary, PhysicalEntity, BioSource, Rna\n",
    "from pybiopax.biopax.model import BiochemicalReaction, Stoichiometry, Provenance, PublicationXref\n",
    "\n",
    "from pybiopax.biopax.model import BioPaxModel\n",
    "from pybiopax.biopax.base import Protein\n",
    "from pybiopax.api import model_to_owl_str"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12a7f37b",
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def create_repressilator(biopax_path: Path):\n",
    "    objects = []\n",
    "    '''\n",
    "    <bp:Protein rdf:about=\"PX\">\n",
    "     <bp:xref rdf:resource=\"http://identifiers.org/uniprot/P03023\" />\n",
    "     <bp:displayName rdf:datatype = \"http://www.w3.org/2001/XMLSchema#string\">LacI protein</bp:displayName>\n",
    "    </bp:Protein>\n",
    "    \n",
    "    Protein(LacI protein)\n",
    "    {'_controller_of': set(), 'xref': [<pybiopax.biopax.util.UnificationXref object at 0x7fe1eb423160>], \n",
    "    'display_name': 'LacI protein', \n",
    "    'standard_name': None, \n",
    "    '_name': [], 'evidence': [], \n",
    "    'uid': 'PX', 'comment': [], 'availability': None, 'data_source': [], \n",
    "    '_participant_of': set(), 'feature': [], \n",
    "    'not_feature': [], 'member_physical_entity': [], \n",
    "    'cellular_location': None, '_component_of': set(), \n",
    "    '_member_physical_entity_of': {PhysicalEntity(PX), PhysicalEntity(PX), PhysicalEntity(PX)}, 'entity_reference': None}\n",
    "    \n",
    "    '''\n",
    "    # p1 = Protein(uid=\"PX\", display_name=\"LacI protein\")\n",
    "    # objects.append(p1)\n",
    "\n",
    "    '''\n",
    "    <bp:PhysicalEntity rdf:about=\"RIGHT_0_conversion_Reaction5_PY\">\n",
    "     <bp:cellularLocation rdf:resource=\"cell\" />\n",
    "     <bp:displayName rdf:datatype = \"http://www.w3.org/2001/XMLSchema#string\">PY</bp:displayName>\n",
    "     <bp:memberPhysicalEntity rdf:resource=\"PY\" />\n",
    "    </bp:PhysicalEntity>\n",
    "    \n",
    "    '''\n",
    "\n",
    "    p1 = PhysicalEntity(uid=\"RIGHT_0_conversion_Reaction5_PY\", display_name=\"PY\", cellular_location=\"cell\", member_physical_entity=\"PY\")\n",
    "    objects.append(p1)\n",
    "\n",
    "    # '''\n",
    "    # <bp:Stoichiometry rdf:about=\"LEFT_0_conversion_Reaction2_Y_STOICHIOMETRY\">\n",
    "    #  <bp:physicalEntity rdf:resource=\"LEFT_0_conversion_Reaction2_Y\" />\n",
    "    #  <bp:stoichiometricCoefficient rdf:datatype = \"http://www.w3.org/2001/XMLSchema#float\">1.0</bp:stoichiometricCoefficient>\n",
    "    # </bp:Stoichiometry>\n",
    "    # '''\n",
    "    s1 = Stoichiometry(uid=\"LEFT_0_conversion_Reaction2_Y_STOICHIOMETRY\", physical_entity=\"LEFT_0_conversion_Reaction2_Y\", stoichiometric_coefficient=\"1\")\n",
    "    objects.append(s1)\n",
    "\n",
    "    # '''\n",
    "    # <bp:Provenance rdf:about=\"datasource_1\">\n",
    "    #  <bp:xref rdf:resource=\"http://identifiers.org/biomodels.db/BIOMD0000000012\" />\n",
    "    #  <bp:displayName rdf:datatype = \"http://www.w3.org/2001/XMLSchema#string\">BioModels Database</bp:displayName>\n",
    "    # </bp:Provenance>\n",
    "    # '''\n",
    "    p3 = Provenance(uid=\"datasource_1\", xref=\"http://identifiers.org/biomodels.db/BIOMD0000000012\", display_name=\"BioModels Database\")\n",
    "    objects.append(p3)\n",
    "    #\n",
    "    p4 = PublicationXref(uid=\"http://identifiers.org/pubmed/10659856\", id=\"10659856\", db=\"PubMed\")\n",
    "    objects.append(p4)\n",
    "\n",
    "    p5 = Provenance(uid=\"datasource_2\", xref=\"http://identifiers.org/biomodels.db/MODEL6615351360\", display_name=\"BioModels Database\")\n",
    "    objects.append(p5)\n",
    "\n",
    "    u1 = UnificationXref(uid=\"http://identifiers.org/biomodels.db/BIOMD0000000012\", id=\"BIOMD0000000012\", db=\"BioModels Database\")\n",
    "    objects.append(u1)\n",
    "\n",
    "    u2 = UnificationXref(uid=\"http://identifiers.org/obo.go/GO:0006402\", id=\"GO:0006402\", db=\"Gene Ontology\")\n",
    "    objects.append(u2)\n",
    "\n",
    "    s2 = Stoichiometry(uid=\"LEFT_0_conversion_Reaction7_PX_STOICHIOMETRY\", physical_entity=\"LEFT_0_conversion_Reaction7_PX\", stoichiometric_coefficient=\"1\")\n",
    "    objects.append(s2)\n",
    "\n",
    "\n",
    "    #\n",
    "    # '''\n",
    "    # <bp:Control rdf:about=\"control_Reaction4_X_0\">\n",
    "    #  <bp:controlled rdf:resource=\"conversion_Reaction4\" />\n",
    "    #  <bp:controlType rdf:datatype = \"http://www.w3.org/2001/XMLSchema#string\">INHIBITION</bp:controlType>\n",
    "    #  <bp:controller rdf:resource=\"PEP0_control_Reaction4X\" />\n",
    "    # </bp:Control>\n",
    "    # '''\n",
    "    c1 = Control(uid=\"control_Reaction4_X_0\", controlled=\"conversion_Reaction4\", control_type=\"INHIBITION\", controller=\"PEP0_control_Reaction4X\")\n",
    "    objects.append(c1)\n",
    "\n",
    "    s2 = Stoichiometry(uid=\"RIGHT_0_conversion_Reaction6_PZ_STOICHIOMETRY\", physical_entity=\"RIGHT_0_conversion_Reaction6_PZ\", stoichiometric_coefficient=\"1\")\n",
    "    objects.append(s2)\n",
    "\n",
    "    c2 = PhysicalEntity(uid=\"PEP0_control_Reaction4X\", cellular_location=\"cell\", display_name=\"X\", member_physical_entity=\"X\")\n",
    "    objects.append(c2)\n",
    "\n",
    "    u3 = UnificationXref(uid=\"http://identifiers.org/obo.go/GO:0030163\", id=\"GO:0030163\", db=\"Gene Ontology\")\n",
    "    objects.append(u3)\n",
    "\n",
    "    s3 = Stoichiometry(uid=\"RIGHT_0_conversion_Reaction12_Z_STOICHIOMETRY\", physical_entity=\"RIGHT_0_conversion_Reaction12_Z\", stoichiometric_coefficient=\"1\")\n",
    "    objects.append(s3)\n",
    "\n",
    "    c3 = PhysicalEntity(uid=\"LEFT_0_conversion_Reaction7_PX\", cellular_location=\"cell\", display_name=\"PX\", member_physical_entity=\"PX\")\n",
    "    objects.append(c3)\n",
    "\n",
    "    s4 = Stoichiometry(uid=\"RIGHT_0_conversion_Reaction11_Y_STOICHIOMETRY\", physical_entity=\"RIGHT_0_conversion_Reaction11_Y\", stoichiometric_coefficient=\"1\")\n",
    "    objects.append(s4)\n",
    "\n",
    "    c4 = CellularLocationVocabulary(uid=\"cell\", term=\"cell\", xref=\"http://identifiers.org/obo.go/GO:0005623\")\n",
    "    objects.append(c4)\n",
    "\n",
    "    b1 = BiochemicalReaction(uid=\"conversion_Reaction12\", xref=\"http://identifiers.org/obo.go/GO:0006351\", participant_stoichiometry=\"RIGHT_0_conversion_Reaction12_Z_STOICHIOMETRY\", display_name=\"transcription of CI\", right=\"RIGHT_0_conversion_Reaction12_Z\")\n",
    "    objects.append(b1)\n",
    "\n",
    "    b1 = BiochemicalReaction(uid=\"conversion_Reaction11\", xref=\"http://identifiers.org/obo.go/GO:0006351\", participant_stoichiometry=\"RIGHT_0_conversion_Reaction11_Y_STOICHIOMETRY\", display_name=\"transcription of TetR\", right=\"RIGHT_0_conversion_Reaction11_Y\")\n",
    "    objects.append(b1)\n",
    "\n",
    "    b1 = BiochemicalReaction(uid=\"conversion_Reaction10\", xref=\"http://identifiers.org/obo.go/GO:0006351\", participant_stoichiometry=\"RIGHT_0_conversion_Reaction10_X_STOICHIOMETRY\", display_name=\"transcription of LacI\", right=\"RIGHT_0_conversion_Reaction10_X\")\n",
    "    objects.append(b1)\n",
    "\n",
    "    p1 = PhysicalEntity(uid=\"LEFT_0_conversion_Reaction9_PZ\", cellular_location=\"cell\", display_name=\"PZ\", member_physical_entity=\"PZ\")\n",
    "    objects.append(p1)\n",
    "\n",
    "    u1 = UnificationXref(uid=\"http://identifiers.org/kegg.compound/C00046\", id=\"C00046\", db=\"KEGG Compound\")\n",
    "    objects.append(u1)\n",
    "\n",
    "    u1 = UnificationXref(uid=\"http://identifiers.org/obo.go/GO:0005623\", id=\"GO:0005623\", db=\"Gene Ontology\")\n",
    "    objects.append(u1)\n",
    "\n",
    "    p1 = PhysicalEntity(uid=\"LEFT_0_conversion_Reaction8_PY\", cellular_location=\"cell\", display_name=\"PY\", member_physical_entity=\"PY\")\n",
    "    objects.append(p1)\n",
    "\n",
    "    u1 = UnificationXref(uid=\"http://identifiers.org/chebi/CHEBI:33699\", id=\"CHEBI:33699\", db=\"ChEBI\")\n",
    "    objects.append(u1)\n",
    "\n",
    "    p1 = Protein(uid=\"PX\", xref=\"http://identifiers.org/uniprot/P03023\", display_name=\"LacI protein\")\n",
    "    objects.append(p1)\n",
    "\n",
    "    p1 = Protein(uid=\"PY\", xref=\"http://identifiers.org/uniprot/P04483\", display_name=\"TetR protein\")\n",
    "    objects.append(p1)\n",
    "\n",
    "    s1 = Stoichiometry(uid=\"RIGHT_0_conversion_Reaction10_X_STOICHIOMETRY\", physical_entity=\"RIGHT_0_conversion_Reaction10_X\", stoichiometric_coefficient=\"1\")\n",
    "    objects.append(s1)\n",
    "\n",
    "    p1 = Protein(uid=\"PZ\", xref=\"http://identifiers.org/uniprot/P03034\", display_name=\"cI protein\")\n",
    "    objects.append(p1)\n",
    "\n",
    "    c1 = BiochemicalReaction(uid=\"conversion_Reaction2\", xref=\"http://identifiers.org/obo.go/GO:0006402\", participant_stoichiometry=\"LEFT_0_conversion_Reaction2_Y_STOICHIOMETRY\", left=\"LEFT_0_conversion_Reaction2_Y\", display_name=\"degradation of TetR transcripts\")\n",
    "    objects.append(c1)\n",
    "\n",
    "    s1 = Stoichiometry(uid=\"RIGHT_0_conversion_Reaction4_PX_STOICHIOMETRY\", physical_entity=\"RIGHT_0_conversion_Reaction4_PX\", stoichiometric_coefficient=\"1\")\n",
    "    objects.append(s1)\n",
    "\n",
    "    p1 = PhysicalEntity(uid=\"LEFT_0_conversion_Reaction3_Z\", cellular_location=\"cell\", display_name=\"Z\", member_physical_entity=\"Z\")\n",
    "    objects.append(p1)\n",
    "\n",
    "    c1 = BiochemicalReaction(uid=\"conversion_Reaction3\", xref=\"http://identifiers.org/obo.go/GO:0006402\", participant_stoichiometry=\"LEFT_0_conversion_Reaction3_Z_STOICHIOMETRY\", left=\"LEFT_0_conversion_Reaction3_Z\", display_name=\"degradation of CI transcripts\")\n",
    "    objects.append(c1)\n",
    "\n",
    "    c1 = BiochemicalReaction(uid=\"conversion_Reaction4\", xref=\"http://identifiers.org/obo.go/GO:0006412\", participant_stoichiometry=\"RIGHT_0_conversion_Reaction4_PX_STOICHIOMETRY\", display_name=\"translation of LacI\", right= \"RIGHT_0_conversion_Reaction4_PX\")\n",
    "    objects.append(c1)\n",
    "\n",
    "    c1 = BiochemicalReaction(uid=\"conversion_Reaction5\", xref=\"http://identifiers.org/obo.go/GO:0006412\", participant_stoichiometry=\"RIGHT_0_conversion_Reaction5_PY_STOICHIOMETRY\", display_name=\"translation of TetR\", right= \"RIGHT_0_conversion_Reaction5_PY\")\n",
    "    objects.append(c1)\n",
    "\n",
    "    p1 = PhysicalEntity(uid=\"LEFT_0_conversion_Reaction2_Y\", cellular_location=\"cell\", display_name=\"Y\", member_physical_entity=\"Y\")\n",
    "    objects.append(p1)\n",
    "\n",
    "    c1 = BiochemicalReaction(uid=\"conversion_Reaction6\", xref=\"http://identifiers.org/obo.go/GO:0006412\", participant_stoichiometry=\"RIGHT_0_conversion_Reaction6_PZ_STOICHIOMETRY\", display_name=\"translation of CI\", right= \"RIGHT_0_conversion_Reaction6_PZ\")\n",
    "    objects.append(c1)\n",
    "\n",
    "    c1 = BiochemicalReaction(uid=\"conversion_Reaction7\", xref=\"http://identifiers.org/obo.go/GO:0030163\", participant_stoichiometry=\"LEFT_0_conversion_Reaction7_PX_STOICHIOMETRY\", display_name=\"degradation of LacI\", left= \"LEFT_0_conversion_Reaction7_PX\")\n",
    "    objects.append(c1)\n",
    "\n",
    "    s1 = Stoichiometry(uid=\"LEFT_0_conversion_Reaction9_PZ_STOICHIOMETRY\", physical_entity=\"LEFT_0_conversion_Reaction9_PZ\", stoichiometric_coefficient=\"1\")\n",
    "    objects.append(s1)\n",
    "\n",
    "    c1 = BiochemicalReaction(uid=\"conversion_Reaction8\", xref=\"http://identifiers.org/obo.go/GO:0030163\", participant_stoichiometry=\"LEFT_0_conversion_Reaction8_PY_STOICHIOMETRY\", display_name=\"degradation of TetR\", left= \"LEFT_0_conversion_Reaction8_PY\")\n",
    "    objects.append(c1)\n",
    "\n",
    "    c1 = BiochemicalReaction(uid=\"conversion_Reaction9\", xref=\"http://identifiers.org/obo.go/GO:0030163\", participant_stoichiometry=\"LEFT_0_conversion_Reaction9_PZ_STOICHIOMETRY\", display_name=\"degradation of CI\", left= \"LEFT_0_conversion_Reaction9_PZ\")\n",
    "    objects.append(c1)\n",
    "\n",
    "    u1 = UnificationXref(uid=\"http://identifiers.org/taxonomy/562\", id=\"562\", db=\"Taxonomy\")\n",
    "    objects.append(u1)\n",
    "\n",
    "    s1 = Stoichiometry(uid=\"LEFT_0_conversion_Reaction1_X_STOICHIOMETRY\", physical_entity=\"LEFT_0_conversion_Reaction1_X\", stoichiometric_coefficient=\"1\")\n",
    "    objects.append(s1)\n",
    "\n",
    "    p1 = PhysicalEntity(uid=\"RIGHT_0_conversion_Reaction4_PX\", cellular_location=\"cell\", display_name=\"PX\", member_physical_entity=\"PX\")\n",
    "    objects.append(p1)\n",
    "\n",
    "    u1 = UnificationXref(uid=\"http://identifiers.org/uniprot/P03034\", id=\"P03034\", db=\"UniProt Knowledgebase\")\n",
    "    objects.append(u1)\n",
    "\n",
    "    p1 = PhysicalEntity(uid=\"PEP0_control_Reaction10PZ\", cellular_location=\"cell\", display_name=\"PZ\", member_physical_entity=\"PZ\")\n",
    "    objects.append(p1)\n",
    "\n",
    "    s1 = Stoichiometry(uid=\"LEFT_0_conversion_Reaction3_Z_STOICHIOMETRY\", physical_entity=\"LEFT_0_conversion_Reaction3_Z\", stoichiometric_coefficient=\"1\")\n",
    "    objects.append(s1)\n",
    "\n",
    "    p1 = PhysicalEntity(uid=\"RIGHT_0_conversion_Reaction10_X\", cellular_location=\"cell\", display_name=\"X\", member_physical_entity=\"X\")\n",
    "    objects.append(p1)\n",
    "\n",
    "    b1 = BioSource(uid=\"biosource_1\", xref=\"http://identifiers.org/taxonomy/562\")\n",
    "    objects.append(b1)\n",
    "\n",
    "    p1 = PhysicalEntity(uid=\"RIGHT_0_conversion_Reaction11_Y\", cellular_location=\"cell\", display_name=\"Y\", member_physical_entity=\"Y\")\n",
    "    objects.append(p1)\n",
    "\n",
    "    p1 = PhysicalEntity(uid=\"RIGHT_0_conversion_Reaction12_Z\", cellular_location=\"cell\", display_name=\"Z\", member_physical_entity=\"Z\")\n",
    "    objects.append(p1)\n",
    "\n",
    "    s1 = Stoichiometry(uid=\"LEFT_0_conversion_Reaction8_PY_STOICHIOMETRY\", physical_entity=\"LEFT_0_conversion_Reaction8_PY\", stoichiometric_coefficient=\"1\")\n",
    "    objects.append(s1)\n",
    "\n",
    "    p1 = PhysicalEntity(uid=\"RIGHT_0_conversion_Reaction6_PZ\", cellular_location=\"cell\", display_name=\"PZ\", member_physical_entity=\"PZ\")\n",
    "    objects.append(p1)\n",
    "\n",
    "    p1 = PhysicalEntity(uid=\"PEP0_control_Reaction6Z\", cellular_location=\"cell\", display_name=\"Z\", member_physical_entity=\"Z\")\n",
    "    objects.append(p1)\n",
    "\n",
    "    p1 = PhysicalEntity(uid=\"PEP0_control_Reaction11PX\", cellular_location=\"cell\", display_name=\"PX\", member_physical_entity=\"PX\")\n",
    "    objects.append(p1)\n",
    "\n",
    "    r1 = Rna(uid=\"Y\", xref=[\"http://identifiers.org/chebi/CHEBI:33699\",\"http://identifiers.org/kegg.compound/C00046\",\"http://identifiers.org/uniprot/P04483\"], display_name=\"TetR mRNA\")\n",
    "    objects.append(r1)\n",
    "\n",
    "    r1 = Rna(uid=\"X\", xref=[\"http://identifiers.org/chebi/CHEBI:33699\",\"http://identifiers.org/kegg.compound/C00046\",\"http://identifiers.org/uniprot/P03023\"], display_name=\"LacI mRNA\")\n",
    "    objects.append(r1)\n",
    "\n",
    "    s1 = Stoichiometry(uid=\"RIGHT_0_conversion_Reaction5_PY_STOICHIOMETRY\", physical_entity=\"RIGHT_0_conversion_Reaction5_PY\", stoichiometric_coefficient=\"1\")\n",
    "    objects.append(s1)\n",
    "\n",
    "    r1 = Rna(uid=\"Z\", xref=[\"http://identifiers.org/chebi/CHEBI:33699\",\"http://identifiers.org/kegg.compound/C00046\",\"http://identifiers.org/uniprot/P03034\"], display_name=\"cI mRNA\")\n",
    "    objects.append(r1)\n",
    "\n",
    "    c1 = Control(uid=\"control_Reaction6_Z_0\", controlled=\"conversion_Reaction6\", control_type=\"INHIBITION\", controller=\"PEP0_control_Reaction6Z\")\n",
    "    objects.append(c1)\n",
    "\n",
    "    c1 = Control(uid=\"control_Reaction12_PY_0\", controlled=\"conversion_Reaction12\", control_type=\"INHIBITION\", controller=\"PEP0_control_Reaction12PY\")\n",
    "    objects.append(c1)\n",
    "\n",
    "    p1 = PhysicalEntity(uid=\"PEP0_control_Reaction12PY\", cellular_location=\"cell\", display_name=\"PY\", member_physical_entity=\"PY\")\n",
    "    objects.append(p1)\n",
    "\n",
    "    u1 = UnificationXref(uid=\"http://identifiers.org/biomodels.db/MODEL6615351360\", id=\"MODEL6615351360\", db=\"BioModels Database\")\n",
    "    objects.append(u1)\n",
    "\n",
    "    u1 = UnificationXref(uid=\"http://identifiers.org/uniprot/P03023\", id=\"P03023\", db=\"UniProt Knowledgebase\")\n",
    "    objects.append(u1)\n",
    "\n",
    "    c1 = Control(uid=\"control_Reaction10_PZ_0\", controlled=\"conversion_Reaction10\", control_type=\"INHIBITION\", controller=\"PEP0_control_Reaction10PZ\")\n",
    "    objects.append(c1)\n",
    "\n",
    "    u1 = UnificationXref(uid=\"http://identifiers.org/obo.go/GO:0006412\", id=\"GO:0006412\", db=\"Gene Ontology\")\n",
    "    objects.append(u1)\n",
    "\n",
    "    u1 = UnificationXref(uid=\"http://identifiers.org/obo.go/GO:0040029\", id=\"GO:0040029\", db=\"Gene Ontology\")\n",
    "    objects.append(u1)\n",
    "\n",
    "    u1 = UnificationXref(uid=\"http://identifiers.org/obo.go/GO:0006351\", id=\"GO:0006351\", db=\"Gene Ontology\")\n",
    "    objects.append(u1)\n",
    "\n",
    "    c1 = Control(uid=\"control_Reaction5_Y_0\", controlled=\"conversion_Reaction5\", control_type=\"INHIBITION\", controller=\"PEP0_control_Reaction5Y\")\n",
    "    objects.append(c1)\n",
    "\n",
    "    p1 = PhysicalEntity(uid=\"LEFT_0_conversion_Reaction1_X\", cellular_location=\"cell\", display_name=\"X\", member_physical_entity=\"X\")\n",
    "    objects.append(p1)\n",
    "\n",
    "    p1 = PhysicalEntity(uid=\"PEP0_control_Reaction5Y\", cellular_location=\"cell\", display_name=\"Y\", member_physical_entity=\"Y\")\n",
    "    objects.append(p1)\n",
    "\n",
    "    c1 = Control(uid=\"control_Reaction11_PX_0\", controlled=\"conversion_Reaction11\", control_type=\"INHIBITION\", controller=\"PEP0_control_Reaction11PX\")\n",
    "    objects.append(c1)\n",
    "\n",
    "    c1 = BiochemicalReaction(uid=\"conversion_Reaction1\", xref=\"http://identifiers.org/obo.go/GO:0006402\", participant_stoichiometry=\"LEFT_0_conversion_Reaction1_X_STOICHIOMETRY\", display_name=\"degradation of LacI transcripts\", left= \"LEFT_0_conversion_Reaction1_X\")\n",
    "    objects.append(c1)\n",
    "\n",
    "    u1 = UnificationXref(uid=\"http://identifiers.org/uniprot/P04483\", id=\"P04483\", db=\"UniProt Knowledgebase\")\n",
    "    objects.append(u1)\n",
    "    # '''\n",
    "    # <bp:BiochemicalReaction rdf:about=\"conversion_Reaction12\">\n",
    "    #  <bp:xref rdf:resource=\"http://identifiers.org/obo.go/GO:0006351\" />\n",
    "    #  <bp:participantStoichiometry rdf:resource=\"RIGHT_0_conversion_Reaction12_Z_STOICHIOMETRY\" />\n",
    "    #  <bp:displayName rdf:datatype = \"http://www.w3.org/2001/XMLSchema#string\">transcription of CI</bp:displayName>\n",
    "    #  <bp:right rdf:resource=\"RIGHT_0_conversion_Reaction12_Z\" />\n",
    "    # </bp:BiochemicalReaction>\n",
    "    # '''\n",
    "    # b1=BiochemicalReaction(uid=\"conversion_Reaction12\", displayName=\"transcription of CI\")\n",
    "    # objects.append(b1)\n",
    "\n",
    "    model = BioPaxModel(objects={o.uid: o for o in objects}, xml_base=\"http://www.biopax.org/release/biopax-level3.owl#\")\n",
    "    owl_str: str = model_to_owl_str(model)\n",
    "\n",
    "    print(\"-\" * 80)\n",
    "    print(owl_str)\n",
    "    print(\"-\" * 80)\n",
    "\n",
    "    # FIXME: validation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73a779ca",
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "if __name__ == \"__main__\":\n",
    "    # print(model)\n",
    "    RESOURCES_DIR: Path = Path(__file__).parent / \"resources\"\n",
    "    RESULTS_DIR: Path = RESOURCES_DIR / \"results\"\n",
    "    # from combine_notebooks import RESULTS_DIR\n",
    "\n",
    "    create_repressilator(biopax_path=RESULTS_DIR / \"repressilator_biopax.owl\")\n",
    "\n",
    "\n",
    "# if True:\n",
    "#     # Reading a model\n",
    "#     model1 = pybiopax.model_from_owl_file(str(RESOURCES_DIR / \"BIOMD0000000012-biopax3.owl\"))\n",
    "#     # biopax = pybiopax.BiochemicalReaction()\n",
    "#     print(model1)\n",
    "\n",
    "#     for obj_key, obj in model1.objects.items():\n",
    "#         print(\"-\" * 80)\n",
    "#         print(obj_key)\n",
    "#         print(obj)\n",
    "#         print(obj.__dict__)"
   ]
  }
 ],
 "metadata": {},
 "nbformat": 4,
 "nbformat_minor": 5
}
